zrml/prediction-markets/src/tests.rs

```diff
@@ -3431,6 +3431,187 @@ fn on_resolution_correctly_reserves_and_unreserves_bonds_for_permissionless_mark
     });
 }
 
+#[test]
+fn report_reserves_outsider_report_bond_and_unreserved_by_on_resolution() {
+    ExtBuilder::default().build().execute_with(|| {
+        let end = 100;
+        assert_ok!(PredictionMarkets::create_market(
+            Origin::signed(ALICE),
+            BOB,
+            MarketPeriod::Block(0..100),
+            get_deadlines(),
+            gen_metadata(2),
+            MarketCreation::Permissionless,
+            MarketType::Categorical(2),
+            MarketDisputeMechanism::SimpleDisputes,
+            ScoringRule::CPMM,
+        ));
+        // Reserve a sentinel amount to check that we don't unreserve too much.
+        assert_ok!(Balances::reserve_named(
+            &PredictionMarkets::reserve_id(),
+            &CHARLIE,
+            SENTINEL_AMOUNT
+        ));
+        let charlie_balance_before = Balances::free_balance(&CHARLIE);
+        assert_eq!(Balances::reserved_balance(&CHARLIE), SENTINEL_AMOUNT);
+        let market = MarketCommons::market(&0).unwrap();
+        let grace_period = end + market.deadlines.grace_period;
+        let report_at = grace_period + market.deadlines.oracle_duration + 1;
+        run_to_block(report_at);
+
+        assert!(market.bonds.outsider.is_none());
+        assert_ok!(PredictionMarkets::report(
+            Origin::signed(CHARLIE),
+            0,
+            OutcomeReport::Categorical(1)
+        ));
+        let market = MarketCommons::market(&0).unwrap();
+        assert_eq!(market.bonds.outsider, Some(Bond::new(CHARLIE, OutsiderBond::get())));
+        assert_eq!(Balances::reserved_balance(&CHARLIE), SENTINEL_AMOUNT + OutsiderBond::get());
+        assert_eq!(Balances::free_balance(&CHARLIE), charlie_balance_before - OutsiderBond::get());
+        let charlie_balance_before = Balances::free_balance(&CHARLIE);
+        // on_resolution called
+        run_blocks(market.deadlines.dispute_duration);
+        assert_eq!(Balances::reserved_balance(&CHARLIE), SENTINEL_AMOUNT);
+        // Check that the outsider gets the OracleBond together with the OutsiderBond
+        assert_eq!(
+            Balances::free_balance(&CHARLIE),
+            charlie_balance_before + OracleBond::get() + OutsiderBond::get()
+        );
+    });
+}
+
+#[test]
+fn disputed_market_results_in_rewarding_outsider() {
```

Comment:
There are four situations, in case of a dispute:

- Oracle reports and is correct
- Oracle reports and is not correct
- Outsider reports and is correct
- Outsider reports and is incorrect

Do we have tests for all four?

Reply:
In case of an _undisputed_ report, do we have a test for "outsider reports and is incorrect"?

Reply:
- [ ] Oracle reports and is correct
- [ ] Oracle reports and is not correct
- [ ] Outsider reports and is correct
- [ ] Outsider reports and is incorrect

Reply:
Regarding your last question. If nobody disputed, then everybody accepted the outsider report (in case it exists). So the case that it's incorrect is not defined. I think you mean [this case](https://github.com/zeitgeistpm/zeitgeist/blob/eaab9bd17668d3111d49753e21c5971745a66c03/zrml/prediction-markets/src/lib.rs#L2464-L2469). It is covered [here](https://github.com/zeitgeistpm/zeitgeist/blob/eaab9bd17668d3111d49753e21c5971745a66c03/zrml/prediction-markets/src/tests.rs#L3391-L3456).

The case that is actually missing is a disputed market on which the outsider reported and the resolved disputed outcome is the same as the outsider report outcome.

Reply:
> Regarding your last question. If nobody disputed, then everybody accepted the outsider report (in case it exists). So the case that it's incorrect is not defined. I think you mean [this case](https://github.com/zeitgeistpm/zeitgeist/blob/eaab9bd17668d3111d49753e21c5971745a66c03/zrml/prediction-markets/src/lib.rs#L2464-L2469). It is covered [here](https://github.com/zeitgeistpm/zeitgeist/blob/eaab9bd17668d3111d49753e21c5971745a66c03/zrml/prediction-markets/src/tests.rs#L3391-L3456).

I have no idea what I meant. If the report is undisputed, then by definition the outsider report is correct. Sorry, no idea what that was supposed to mean.